### ISSUE TITLE
Directory Tree View with Selection State

### ISSUE THREAD

**Issue by ron3545** [ocornut/imgui#7020]
**Version/Branch of Dear ImGui:**

Version: 1.90
Branch: Docking

**Back-end/Renderer/Compiler/OS**

Back-ends: imgui_impl_Win32.cpp + imgui_impl_dx11.cpp 
Operating System: Windows

**My Issue/Question:**

I was trying to create a directory tree view, more like an explorer similar to vscode, on which when the user selected a file, that part of node should be highlighted until the user clicked another file. I managed to do that with the help of the examples provided on imgui_demo.cpp . However, I'm facing some bugs on which there are times that there are multiple nodes being highlighted especially when there are more nodes being rendered, see the image and video below:

https://github.com/ocornut/imgui/assets/86136180/b43f7696-6f75-4823-b804-8f1c84b702ed

![400008719_721802849385096_4743167620286141559_n](https://github.com/ocornut/imgui/assets/86136180/969307c2-4ead-4a7a-9c39-79a54e73bbaf)

PS. The code was taken from #5137 but made some changes on it by adding selection state whenever a leaf node was pressed.

Here is the part that renders the Tree View:
```
static long unsigned int leaf_node_index  = 0;
static int node_clicked = -1;
// barrowed from ImGui_Demo
static long int selection_mask = (1 << 2);
void RecursivelyDisplayDirectoryNode(DirectoryNode& parentNode, size_t nChildren)
{   
    ImGuiTreeNodeFlags node_flags = ImGuiTreeNodeFlags_SpanFullWidth;
    ImGui::PushID(&parentNode);
    if (parentNode.IsDirectory)
    {
        if (ImGui::TreeNodeEx(parentNode.FileName.c_str(), node_flags))
        {   
            for (DirectoryNode& childNode : parentNode.Children){
                RecursivelyDisplayDirectoryNode(childNode, parentNode.Children.size());
                leaf_node_index++;
            }
            ImGui::TreePop();
        }
    }
    else
    {   
        node_flags |= ImGuiTreeNodeFlags_Leaf | ImGuiTreeNodeFlags_NoTreePushOnOpen;
        if((selection_mask & (1 << leaf_node_index )) !=0)
            node_flags |= ImGuiTreeNodeFlags_Selected;

        ImGui::TreeNodeEx(parentNode.FileName.c_str(), node_flags);
        if(ImGui::IsItemClicked() && !ImGui::IsItemToggledOpen()){
            node_clicked = leaf_node_index ;
        }    
    }
    leaf_node_index++;
    ImGui::PopID();
}

void ImplementDirectoryNode()
{
    ImGui::Dummy(ImVec2(0.0f, 13.05f));
    ImGui::Dummy(ImVec2(6.0f, 13.05f));
    ImGui::SameLine();
    
    ImGui::PushFont(FileTreeFont);
    leaf_node_index = 0;
    if(!project_root_node.FileName.empty() || !project_root_node.FullPath.empty())
        RecursivelyDisplayDirectoryNode(project_root_node, project_root_node.Children.size());

    ImGui::Begin("debug");
        ImGui::Text("selection_mask: "); ImGui::SameLine();
        ImGui::Text(std::to_string(selection_mask ).c_str());

        ImGui::Separator();
        ImGui::Text("leaf_node_index: "); ImGui::SameLine();
        ImGui::Text(std::to_string(leaf_node_index ).c_str());

        ImGui::Separator();
        ImGui::Text("node_clicked: "); ImGui::SameLine();
        ImGui::Text(std::to_string(node_clicked ).c_str());
    ImGui::End();
    
    leaf_node_index  = 0;
    if (node_clicked != -1)
        selection_mask = (1 << node_clicked);
    
    ImGui::PopFont();
}
```

Is there a way to prevent those bug from happening?

**Comment by GamingMinds-DanielC** [comment]
Your selection mask is the problem, a 32 bit integer won't cut it (64 will neither). Shifting an integer for its width in bits or more is undefined behavior. In this case (you shift 1 by `leaf_node_index` which gets too big) the undefined behavior seems to be that the amount to shift is bitwise ANDed with 31 (notice how your additional selection is 32 entries off)). Another compiler could give you other other results, but never the ones you want since a single integer just doesn't hold enough bits.

In short: you need to find a better way to store your selection state, like f.e. `std::vector<bool>`.

**Comment by ocornut** [comment]
The proper way to store a selection would be e.g. `std::set<YourItemID>` or `std::set<ImGuiID>`.
The way we do it in some of the demos is not really scalable and probably not a good demo. Storing indices not a good idea either.

In 1.91 I will expose some more helpers to facilitate this, but if you can use a `std::set` do that.
